ambari-server/src/main/java/org/apache/ambari/server/orm/dao/ServiceDesiredStateDAO.java

```diff
@@ -39,8 +38,16 @@
   Provider<EntityManager> entityManagerProvider;
 
   @RequiresSession
-  public ServiceDesiredStateEntity findByPK(ServiceDesiredStateEntityPK primaryKey) {
-    return entityManagerProvider.get().find(ServiceDesiredStateEntity.class, primaryKey);
+  public ServiceDesiredStateEntity findByServiceId(Long serviceId) {
+    TypedQuery<ServiceDesiredStateEntity> query = entityManagerProvider.get()
+            .createNamedQuery("ServiceDesiredStateByServiceId", ServiceDesiredStateEntity.class);
+    query.setParameter("serviceId", serviceId);
+
+    try {
+      return query.getSingleResult();
+    } catch (NoResultException ignored) {
+      return null;
+    }
```

Comment:
I'm not sure that this change is a good idea. Moving away from a PK for this table, it means that any lookup must hit the database instead of the L1 cache. Why was this necessary?

Reply:
Change was made as there is a way to simplify what constitutes as primary here. Servceid will be unique/primary here.

With my current change of using service_id as primary key, and querying it as : 

>>>>> return entityManagerProvider.get().find(ServiceDesiredStateEntity.class, serviceId);

That will get the query serviced via cache.

Reply:
But your final code here does not do this. Instead, to lookup a ServiceDesiredStateEntity via a NamedQuery. That means anytime you need the state of a service, it's hitting the DB.

Reply:
Sure, I was not clear on my earlier comment. Given that we were discussing on whether I should change PK for the *servicedesiredstate* table or not and given that I had changed it, but with the change the change in PK, I was not using EntityManager. 

So, I had posted the change in comment that if we want to go changing the Pk construct (which I had done earlier), and by using the EntityManager (suggested in comments), we can leverage the caching mechanism. 

So, I have incorporated the EntityManager change now.